omc3/kmod/analysis.py

```diff
@@ -247,49 +253,145 @@ def return_df(magnet1_df, magnet2_df, plane):
+        "IP3" : ["BPMW.4L3", "BPMW.4R3"], 
+        "IP4" : ["BPMWA.A5L4", "BPMWA.A5R4"], 
+        "IP5" : ["BPMSW.1L5", "BPMSW.1R5"],
+        "IP6" : ["BPMSE.4L6", "BPMSA.4R6"], 
+        "IP7" : ["BPMW.4L7", "BPMW.4R7"], 
+        "IP8" : ["BPMSW.1L8", "BPMSW.1R8"], 
+    }
+
+    # Selecting BPMs
+    BPML = BPM_dict[kmod_input_params.interaction_point.upper()][0] + '.' + kmod_input_params.beam.upper()
+    BPMR = BPM_dict[kmod_input_params.interaction_point.upper()][1] + '.' + kmod_input_params.beam.upper()
+
+    return BPML, BPMR
+
+def get_BPM_distance(kmod_input_params,BPML,BPMR):
+
+    model_filename = 'twiss_' + kmod_input_params.beam.lower() + '.dat'
+
+    # Position s of BPML
+    if (os.path.exists(os.path.join(f'{kmod_input_params.model_dir}', f'{model_filename}'))): # distance between actual BPMs if model is provided (more precise)
+        twiss_df = tfs.read(os.path.join(f'{kmod_input_params.model_dir}', f'{model_filename}'), index='NAME')
+        pos_1L = twiss_df.loc[BPML,'S']
+        pos_1R = twiss_df.loc[BPMR,'S']
+        BPM_distance = np.abs(pos_1R - pos_1L)/2.0
+    else:
+        BPM_distance = 22.965 # Distance between last quadrupole and IP (LSTAR) (less precise)
+
+    return BPM_distance
+
+def get_phase_from_model(kmod_input_params,plane):        # get phase from twiss model
+    model_filename = 'twiss_' + kmod_input_params.beam + '.dat'
+    twiss_df = tfs.read(os.path.join(f'{kmod_input_params.model_dir}', f'{model_filename}'), index='NAME')
+    BPML, BPMR = get_BPM(kmod_input_params)[0], get_BPM(kmod_input_params)[1]
+    phase_1L = twiss_df.loc[BPML,'MU'+plane]
+    phase_1R = twiss_df.loc[BPMR,'MU'+plane]
+    phase_adv_model = abs(phase_1R - phase_1L)
+    phase_adv_err = 0.5e-3 # this number is given by Andrea's estimations 
+
+    return phase_adv_model, phase_adv_err
+
+def get_phase_from_measurement(kmod_input_params,plane):
+    # get measured phase from getphase[x/y].out
+    # phase_df = tfs.read( os.path.join(f'{kmod_input_params.measurement_dir}',f'getphase{plane.lower()}.out'), index='NAME')
+    ## get measured phase from phase_{x/y}.out
+    phase_df = tfs.read( os.path.join(f'{kmod_input_params.measurement_dir}',f'phase_{plane.lower()}.out'), index='NAME')
+    BPML,BPMR = get_BPM(kmod_input_params)[0], get_BPM(kmod_input_params)[1]
+    phase_adv_model = phase_df.loc[BPML,'PHASE'+plane]
+    phase_adv_err = phase_df.loc[BPML,'STDPH'+plane]
+    # getphase is python2, python3 is phase_x/y
+    return phase_adv_model, phase_adv_err
+
+def phase_constraint(kmod_input_params,plane):
+
+    # weight given to phase
+    weight = kmod_input_params.phase_weight
+
+    model_filename = 'twiss_' + kmod_input_params.beam + '.tfs'
+
+    # if measured data exists
+    #if os.path.exists(os.path.join(f'{kmod_input_params.measurement_dir}',f'getphase{plane.lower()}.out')):
+    if os.path.exists(os.path.join(f'{kmod_input_params.measurement_dir}',f'phase_{plane.lower()}.out')):
```

Comment:
file name root and suffix( which is .tfs) are in omc3/optics_measurements/constants.py